### PR TITLE
Take gradient accumulation into account when defining samplers

### LINKED ISSUE #14638: Gradient accumulation causing different training curves
For a pretraining experiment using the built-in `Trainer`, setting batch size 32 x 16 accumulation steps seems to yield a different training loss curve from 64 x 8. Even though they converge to approximately the same value, shouldn't the curves be exactly the same? What would cause the difference?
image

I've also seen similar things with using 1 vs. multiple GPUs (under DataParallel), which may or may not be relevant.

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR takes the gradient accumulation steps into account when defining samplers that use the batch size (like the `LengthGroupedSampler`) so that training with large batch size or training with smaller batch size and gradient accumulation (e.g. batch size 64 or batch size 8 and gradient accumulation steps of 8) yield the same results.

Fixes #14638

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
Good for me!
